presto-native-execution/presto_cpp/main/PrestoServer.cpp

```diff
@@ -649,6 +659,11 @@ PrestoServer::getHttpServerFilters() {
             httpServer_.get()));
   }
 
+  // Always add the autnetication filter to make sure the worker configuration
+  // is in line with the overall cluster configuration e.g. cannot have a worker
+  // without JWT enabled.
```

Comment:
@nmahadevuni @tdcmeehan from the Java implementation https://github.com/prestodb/presto/pull/19706/files, we understood that the authentication filter is always added even if JWT is disabled. This is to check if the HTTP headers have the presto-internal-bearer and throw an error if JWT is disabled. Did we get this right?

Reply:
Authentication filter has a list of authenticators that will be applied to all incoming requests. The list of authenticators can be empty if we didn't configure any authentication though. This is not specific to JWT. There are other authenticators PASSWORD,CERTIFICATE etc., When JWT is enabled, the request will have the header "X-Presto-Internal-Bearer". In effect, this is also one of the authenticators. If none is present, then the request is passed through, but if one or more authenticators are present, at least one should be successful for each request to go through.

Reply:
" If none is present, then the request is passed through"
I think this is a problem. 

Looking at the Java code then it only validates internal requests if the presto bearer is set. If not, then it falls through - and if not further authenticators are available the request passes through and is processed.

(in doFilter
```
        if (internalAuthenticationManager.isInternalRequest(request)) {. <--- presto bearer present or not
            Principal principal = internalAuthenticationManager.authenticateInternalRequest(request);
            if (principal == null) {
                response.sendError(SC_UNAUTHORIZED);
                return;
            }
            nextFilter.doFilter(withPrincipal(request, principal), response);
            return;
        }
```
)

As a result, even if the server is configured to validate (JWT) requests, an attacker sending a simple request without the bearer passes authentication. This doesn't seem right.

In the C++ implementation the bearer is required if worker has JWT authentication configured.
It also rejects a request if the bearer is present and the worker is not configured for JWT authentication.

This hardens the worker and cluster config. I think on the Java side there is a security leak.
We should add similar behavior to Java.

Reply:
CC: @tdcmeehan

Reply:
The checks for bearer and JWT config can be hardened only on workers since they don't need to worry about external authentication. In coordinator(where it can a worker too), we cannot know upfront if the request is internal or external. So, if we harden these checks for internal JWT, then we will fail all external requests.